### ISSUE TITLE
@exception description paragraph does not accept @ref tag

### ISSUE THREAD

**Issue by MarkShackman** [doxygen/doxygen#9831]
Both @param and @exception say that their description "paragraph has no special internal structure. All visual enhancement commands may be used inside the paragraph."

However, @exception descriptions don't accept @ref commands, whereas @param descriptions do:

```
## @file throwsBug

##
# Example class
class Example():

  def function(p):
    """ Example function
    """
    pass

  ## Documentation using doxygen style comment
  # @param p @ref throwsBug.Example.function() "function()"
  # @throws @ref throwsBug.Example.function() "function()"
  def useDoxygenStyle(p):
    pass
```
Using doxygen 1.9.6 with an unchanged Doxyfile.

**Comment by albert-github** [comment]
The documentation of `\throws` gives:

> 24.103 \throws <exception-object> { exception description }


and as \`throws` is synonymous with `\exception`:

> 24.79 \exception <exception-object> { exception description }
>
> Starts an exception description for an exception object with name <exception-object>. Followed by a description
of the exception. The existence of the exception object is not checked. The text of the paragraph has no special
internal structure. All visual enhancement commands may be used inside the paragraph. Multiple adjacent
\exception commands will be joined into a single paragraph. Each exception description will start on a new
line. The \exception description ends when a blank line or some other sectioning command is encountered.
See section \fn for an example.

I think the problem is that you forgot to mention the `<exception-object>` with the `\param` we can see that you used `p` as `<parameter-name>`.

Adding the `<exception-object>` will solve the problem.

I think the problem is usage.

**Comment by MarkShackman** [comment]
Ah, true.
Perhaps refactoring has tightened up on syntax, as the following is working for me under doxygen 1.8.18:
```
   @throws [@ref exceptions.SLException.SLException "SLException"] if unable to communicate
```
The 1.8.18 build uses an edited doxyfile with doxypy.py as a filter, so I'll have to investigate further....

**Comment by albert-github** [comment]
When I use your syntax from https://github.com/doxygen/doxygen/issues/9831#issuecomment-1414106515 in 1.8.18 I get (and a warning):

```
throwsBug.py:14: warning: unexpected token in comment block while parsing the argument of command throws
```

![image](https://user-images.githubusercontent.com/5223533/216399865-c64cedd1-75f4-43af-a500-b0b325f89bf3.png)


so likely that filter does something with the `@exception`

**Comment by MarkShackman** [comment]
I've got a bit further. There's an ALIASES defined as ` "throws=@par Exception^^" ` which changes `@throws` to a standard paragraph, enabling the exception object to be a `@ref`. This works correctly in 1.8.18 with the doxypy.py filter.

However, now that `PYTHON_DOCSTRING` is available, I've removed the filter (so I can replace it with another that hides protected elements) but it seems that the ALIAS inside a Python docstring isn't substituted.

With the doxyfile:
```
ALIASES  = "throws=@par ExceptionsXX^^" \
           "throwz=@par ExceptionsXX^^" \
           "CustomException=ThrowsRefBug.CustomException"
PYTHON_DOCSTRING  = NO
```
the following code shows the problem; the `ALIASES` in the doxygen-style comments are substituted, but those in the docstring aren't:
```
## @file

##
# Custom exception class.
#
class CustomException(Exception):
    ##
    # Get the error code of the exception.
    #
    # @return [int] error code of the exception
    #
    def getErrorCode(self):
        return self._errorCode

  ##
  # \throwz [\ref ThrowsRefBug.CustomException "CustomException"]  if there's a custom problem
  # \throwz [\ref \CustomException "CustomException"]  if there's a custom problem
  def Example():
   """ Documentation

   @throws [\ref \CustomException "CustomException"]  if there's a custom problem

   \throwz [\ref \CustomException "CustomException"]  if there's a custom problem
   """
   pass
```
Have I done something wrong with the docstring?

**Comment by albert-github** [comment]
I added a test function:
```
##
# \throwz [\ref ThrowsRefBug.CustomException "CustomException"]  if there's a custom problem
# \throwz [\ref \CustomException "CustomException"]  if there's a custom problem
def Example1():
 """! Documentation

 @throws [\ref \CustomException "CustomException"]  if there's a custom problem

 \throwz [\ref \CustomException "CustomException"]  if there's a custom problem
 """
 pass
```

and this gives, like one wants (I think) to have:

![image](https://user-images.githubusercontent.com/5223533/216614233-d92464cc-d0d3-4291-b0bd-514962564bd9.png)


It looks like that the comment converter (`commentcnv.l`) does not handle the `"""` correctly in case of `PYTHON_DOCSTRING=YES` (the word `PYTHON_DOCSTRING` doesn't appear at all in the comment converter) and in the comment converter the `ALIASES` are handled.
So this looks like that some handling of the `"""` in case of `PYTHON_DOCSTRING=YES` has to be added to the comment converter.

**Comment by albert-github** [comment]
I've just pushed a proposed patch, pull request #9833

**Comment by MarkShackman** [comment]
Thanks - can confirm that adding the ! works (well, mostly - if the docstring comments are indented by four spaces, they are interpreted as pre-formatted text and appear as a code block; is there any way to fix this?)

**Comment by albert-github** [comment]
Not really, this is intended as in Markdown 
> Code Blocks
>
> Preformatted verbatim blocks can be created by indenting each line in a block of text by at least 4 extra spaces 

(see https://www.doxygen.nl/manual/markdown.html#md_codeblock)

You could disable Markdown (`MARKDOWN_SUPPORT=NO`) but you will loose a lot of nice features.

Btw. Did you try it based on doxygen version 1.9.6 or based on the current master (1.9.7 (f40290be22990ddc164736c5a43352cb82cb5e29))?

**Comment by MarkShackman** [comment]
At a quick glance, the code block effect only seems to occur only where there are blank lines in the docstrings, but removing all blank lines generates even more errors. It's a problem, as Python relies on indentation.

I've not tried the current master (can only use official releases for our builds) and really don't want to lose Markdown support! Will investigate further next week. Thanks for the support so far.

**Comment by albert-github** [comment]
Unfortunately indeed Python relies on indentation.

Can you show the 2 examples you mentioned with and without the blank lines?

**Comment by albert-github** [comment]
Code has been integrated in master on GitHub (please don't close the issue as this will be done at the moment of an official release).

**Comment by MarkShackman** [comment]
Thanks.

[For testing purposes, is there a specific location I can download the built executables from GitHub, or must I start to build locally/wait for an official release?]

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.7.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
